Store shower information in the NUR header

MijnheerD opened the pull request:
Issue being addressed:

Currently no information about the showers (such as the SimShower for simulations) is stored in the header. This makes it cumbersome to check the basic properties of the event for simulations, as typically parameters such as the energy are only stored in the Shower object.

Did you remember to: 
- [ ] include doc strings in all commits ?
- [ ] edit the changelog.txt file describing your changes?

MijnheerD committed:
Add showers entry to header dict

fschlueter commented:
@MijnheerD are you planning to continue working on this?

galaxyvita commented:
I don't mind doing this, if I understand what needs to happen.

fschlueter commented:
> I don't mind doing this, if I understand what needs to happen.

have a look at the existing code and how we copy all station parameters into a dict which is stored in the header. It should be rather straight forward to do the same for all (sim)showers

galaxyvita commented:
Yes, but I looked for existing code which does the same and couldn't find anything. So I'm a buit confused.

galaxyvita commented:
Oke so I think I updated the function in a way that would work, but I don't have an example to run it on.

Commit:
I updated the function adding the shower parameters.

fschlueter committed:
Merge branch 'develop' into add_shower_to_header

fschlueter commented:
left one little comment, otherwise looks fine

Commit:
changed the index to shower.get_id() so it represents something and isn't just 0,1,2,3

Commit:
Temp test

Commit:
removes print statments for testing

fschlueter commented:
oh sorry, that slipped my attention.

fschlueter committed:
Merge branch 'develop' into add_shower_to_header

fschlueter commented on NuRadioReco/modules/io/eventWriter.py:
```diff
@@ -20,6 +20,15 @@ def get_header(evt):
+    for shower in evt.get_showers():
+        sh_id = shower.get_id()
+        header['showers'][sh_id] = shower.get_parameters().copy()
+        if hasattr(shower, "has_sim_shower") and shower.has_sim_shower():
```
i am confused, only the event call has the function `has_sim_shower`?

fschlueter replied:
@galaxyvita this if condition will be always 0, hence you will never store sim_shower data. My suggestion is that you include in the header also a key `sim_showers` and loop over all sim_showers.

galaxyvita commented:
Oke i found that i updated the “read_lofar_data.ipynb” in examples and printed some of the headers there.

Commit:
Fixed header mistake

very stupid...

Commit:
typo

fschlueter commented on NuRadioReco/modules/io/NuRadioRecoio.py:
```diff
@@ -171,26 +171,25 @@ def close_files(self):
     def get_filenames(self):
         return self._filenames
-
+                        
```
meh

fschlueter commented on NuRadioReco/modules/io/NuRadioRecoio.py:
```diff
@@ -171,26 +171,25 @@ def close_files(self):
+        
         self.__event_ids.append(evt_header['event_id'])
-
+    
```
meh ... also in the following lines

fschlueter commented on NuRadioReco/modules/io/NuRadioRecoio.py:
```diff
@@ -201,10 +200,29 @@ def _parse_event_header(self, evt_header):
+
+        if 'showers' in evt_header:
+            if not hasattr(self, '_event_showers'):
+                self._event_showers = {}
```
I have not digested the code fully, but is there a reason why you are not initializing (and importantly resetting) `self._event_showers` similar to `__event_headers` in the function `openFile`?

galaxyvita replied:
Because I did not think about that. This would probably keep the previous files in memory.

fschlueter commented on NuRadioReco/modules/io/NuRadioRecoio.py:
```diff
@@ -201,10 +200,29 @@ def _parse_event_header(self, evt_header):
+                    if key not in self._event_showers[shower_id]:
+                        self._event_showers[shower_id][key] = []
+    
+                    self._event_showers[shower_id][key].append(value)
```
I think this motivated by the structure in the same code for the stations. I do not think that makes sense here (although I am happy to discuss that). I think the idea for stations is to store the information per station accumulating information over all events. For showers, I think it would make more sense to store them per event. Does that make sense to you?

galaxyvita replied:
This does make sense.

fschlueter reviewed (changes requested):
Hey @galaxyvita, unfortunately I found a couple of things we need to fix

galaxyvita committed:
Update NuRadioRecoio.py

Updated code based on Felix comment where would the previous files would be kept in memory because i did not reselt the self._event_showers = {} in the Openfile().

galaxyvita committed:
Update NuRadioRecoio.py

Based on felix’s comments now the shower info is stored on an event by event basis.

Commit:
Based on felix comments refactored the get_sim_shower and added a few tabs

Header now has header = {'stations': {}, 'showers': {}, 'sim_showers':
{}}

galaxyvita committed:
Update eventWriter.py

galaxyvita committed:
Update NuRadioRecoio.py

fschlueter committed:
Add __event_showers attribute to NuRadioRecoio

fschlueter committed:
Refactor NuRadioRecoio
